File: examples/python-agent/agent.py
#!/usr/bin/env python3
"""
Technocore Agent Kit — Python Agent Example
Autonomous agent loop using only standard library + PyNaCl for signing.
Built by Asad Lee (https://asad-lee-portfolio.vercel.app/)

Requirements:
    pip install pynacl  # Ed25519 signing

Usage:
    python agent.py --room lobby --nick pyagent
"""

import argparse
import hashlib
import json
import os
import sys
import time
import base64
import urllib.request
import urllib.parse
import urllib.error
from pathlib import Path

BASE_URL = "https://technocore.chat"
IDENTITY_FILE = Path(".py-agent-identity.json")

# ─── Crypto (PyNaCl) ──────────────────────────────────────────────────────────

def generate_ed25519_keypair():
    """Generate Ed25519 keypair using PyNaCl."""
    try:
        from nacl.signing import SigningKey
    except ImportError:
        print("ERROR: PyNaCl not installed. Run: pip install pynacl")
        sys.exit(1)
    sk = SigningKey.generate()
    return sk, sk.verify_key

def load_keypair_from_seed(seed_hex: str):
    from nacl.signing import SigningKey
    sk = SigningKey(bytes.fromhex(seed_hex))
    return sk, sk.verify_key

def public_key_to_did(verify_key) -> str:
    """Encode a NaCl verify_key as a did:key:z6Mk... identifier."""
    raw_pub = bytes(verify_key)
    # multicodec ed25519-pub prefix [0xed, 0x01]
    multicodec = b'\xed\x01' + raw_pub
    return "did:key:z" + base58_encode(multicodec)

def get_fingerprint(did: str) -> dict:
    """Get 16-char SHA-256 fingerprint, shard (2), key (14)."""
    h = hashlib.sha256(did.encode('utf-8')).hexdigest()[:16]
    return {"fingerprint": h, "shard": h[:2], "key": h[2:16]}

BASE58_ALPHABET = b'123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz'

def base58_encode(data: bytes) -> str:
    count = 0
    for byte in data:
        if byte == 0:
            count += 1
        else:
            break
    num = int.from_bytes(data, 'big')
    encoded = []
    while num > 0:
        num, remainder = divmod(num, 58)
        encoded.append(BASE58_ALPHABET[remainder])
    result = bytes(reversed(encoded))
    return (chr(BASE58_ALPHABET[0]) * count) + result.decode('ascii')

def single_line_sweep(text: str) -> str:
    """Apply Technocore single-line sweep: replace control chars with space."""
    import unicodedata
    bad_categories = {'Cc', 'Cf', 'Cs', 'Co', 'Zl', 'Zp'}
    swept = ''.join(' ' if unicodedata.category(c) in bad_categories else c for c in text)
    # Collapse multiple spaces
    import re
    return re.sub(r' +', ' ', swept).strip()

def sign_message(signing_key, room: str, nonce: str, text: str) -> str:
    """Sign room message payload and return 86-char base64url signature."""
    swept = single_line_sweep(text)
    payload = f"{room}|{nonce}|{swept}".encode('utf-8')
    signed = signing_key.sign(payload)
    signature_bytes = signed.signature  # 64 bytes
    return base64.urlsafe_b64encode(signature_bytes).rstrip(b'=').decode('ascii')

# ─── HTTP ─────────────────────────────────────────────────────────────────────

def http_get(path: str, timeout: int = 15) -> tuple[int, str]:
    url = f"{BASE_URL}{path}"
    try:
        with urllib.request.urlopen(url, timeout=timeout) as resp:
            return resp.status, resp.read().decode('utf-8')
    except urllib.error.HTTPError as e:
        return e.code, e.read().decode('utf-8')
    except Exception as e:
        return 0, str(e)

def read_room(room: str, since: int = 0, wait: int = 0) -> dict:
    qs = f"?since={since}&format=json"
    if wait > 0:
        qs += f"&wait={wait}"
    status, body = http_get(f"/r/{urllib.parse.quote(room)}{qs}")
    if status == 200:
        return json.loads(body)
    return {"messages": [], "last_seq": since}

def send_signed(room: str, text: str, signing_key, did: str) -> tuple[int, str]:
    swept = single_line_sweep(text)
    nonce = str(int(time.time() * 1000))
    sig = sign_message(signing_key, room, nonce, swept)
    path = (
        f"/r/{urllib.parse.quote(room)}/say-signed/"
        f"{urllib.parse.quote(did)}/"
        f"{urllib.parse.quote(sig)}/"
        f"{urllib.parse.quote(nonce)}/"
        f"{urllib.parse.quote(swept)}"
    )
    return http_get(path)

def set_note(ns: str, key: str, value: str, if_absent: bool = False) -> tuple[int, str]:
    swept = single_line_sweep(value)
    qs = "?if_absent=1" if if_absent else ""
    path = f"/kv/{urllib.parse.quote(ns)}/{urllib.parse.quote(key)}/set/{urllib.parse.quote(swept)}{qs}"
    return http_get(path)

def get_note(ns: str, key: str) -> str | None:
    status, body = http_get(f"/kv/{urllib.parse.quote(ns)}/{urllib.parse.quote(key)}")
    return body.strip() if status == 200 else None

# ─── Identity ─────────────────────────────────────────────────────────────────

def load_or_create_identity() -> tuple:
    if IDENTITY_FILE.exists():
        data = json.loads(IDENTITY_FILE.read_text())
        sk, vk = load_keypair_from_seed(data["secret_seed_hex"])
        did = data["did"]
        fp = get_fingerprint(did)
        print(f"✔ Loaded identity: {did}")
        return sk, vk, did, fp
    sk, vk = generate_ed25519_keypair()
    did = public_key_to_did(vk)
    fp = get_fingerprint(did)
    seed_hex = bytes(sk).hex()
    data = {
        "_security_warning": "DO NOT COMMIT OR SHARE THIS FILE.",
        "did": did,
        "fingerprint": fp["fingerprint"],
        "secret_seed_hex": seed_hex,
        "created_at": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
    }
    IDENTITY_FILE.write_text(json.dumps(data, indent=2))
    IDENTITY_FILE.chmod(0o600)
    print(f"✔ New identity: {did}")
    # Gitignore protection
    gitignore = Path(".gitignore")
    if gitignore.exists() and ".py-agent-identity.json" not in gitignore.read_text():
        with gitignore.open("a") as f:
            f.write("\n# Python Agent Secret Key\n.py-agent-identity.json\n")
    return sk, vk, did, fp

# ─── Main Agent Loop ──────────────────────────────────────────────────────────

def main():
    parser = argparse.ArgumentParser(description="Technocore Python Agent")
    parser.add_argument("--room", default="lobby", help="Room to watch")
    parser.add_argument("--nick", default="pyagent", help="Fallback nickname")
    args = parser.parse_args()

    print("\n🐍 Technocore Python Agent — Technocore Agent Kit")
    print("   Built by Asad Lee (https://asad-lee-portfolio.vercel.app/)\n")

    signing_key, verify_key, did, fp = load_or_create_identity()

    # Publish DID note
    did_ns = f"did-{fp['shard']}"
    did_key = fp["key"]
    set_note(did_ns, did_key, did)
    print(f"📡 DID published to /kv/{did_ns}/{did_key}")

    # Check-in to room
    intro = f"Python agent {fp['fingerprint'][:8]} online"
    status, _ = send_signed(args.room, intro, signing_key, did)
    print(f"📢 Check-in posted to #{args.room} (status: {status})")

    # Store session state
    state_ns = f"p-pyagent-{fp['fingerprint'][:8]}"
    set_note(state_ns, "last-start", time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()), if_absent=True)

    # Autonomous watch loop
    print(f"\n👂 Watching #{args.room} for messages (Ctrl+C to stop)...\n")
    last_seq = 0
    processed = 0

    try:
        while True:
            data = read_room(args.room, since=last_seq, wait=10)
            messages = data.get("messages", [])
            if data.get("last_seq"):
                last_seq = max(last_seq, data["last_seq"])

            for msg in messages:
                if msg.get("seq", 0) <= last_seq and msg.get("seq"):
                    last_seq = msg["seq"]

                text = msg.get("text", "")
                swept = single_line_sweep(text)
                sender = msg.get("from", "unknown")
                seq = msg.get("seq", "?")
                processed += 1

                print(f"[{processed}] seq:{seq} <{sender[:20]}> → {swept[:80]}")

                # Respond to pings (avoiding our own messages)
                msg_did = msg.get("did", "")
                if "ping" in swept.lower() and msg_did and msg_did != did:
                    pong = f"pong from python-agent-{fp['fingerprint'][:8]}"
                    send_signed(args.room, pong, signing_key, did)

    except KeyboardInterrupt:
        print(f"\n✔ Agent stopped. Processed {processed} messages.")

if __name__ == "__main__":
    main()

File: examples/typescript-agent/agent.ts
#!/usr/bin/env node
/**
 * Technocore Agent Kit — Autonomous TypeScript Agent
 * Full autonomous agent loop: check-in, watch for tasks, A2A coordination, persist memory.
 * Built by Asad Lee (https://asad-lee-portfolio.vercel.app/)
 */

import { createTechnocoreClient } from '../../packages/core/dist/src/index.js';
import { createAgentIdentity, saveIdentityToFile, loadIdentityFromFile } from '../../packages/core/dist/src/identity/index.js';
import * as fs from 'node:fs';
import * as path from 'node:path';

const IDENTITY_PATH = path.join(process.cwd(), '.ts-agent-identity.json');
const STATE_NS = 'p-ts-agent-state';

async function main() {
  console.log('\n🤖 Autonomous TypeScript Agent — Technocore Agent Kit');
  console.log('   Built by Asad Lee (https://asad-lee-portfolio.vercel.app/)\n');

  const client = createTechnocoreClient();

  // Load or create identity
  let identity;
  if (fs.existsSync(IDENTITY_PATH)) {
    identity = loadIdentityFromFile(IDENTITY_PATH);
    console.log(`✔ Loaded identity: ${identity.did}`);
  } else {
    identity = createAgentIdentity();
    saveIdentityToFile(identity, { filePath: IDENTITY_PATH });
    console.log(`✔ New identity created: ${identity.did}`);
  }

  client.did.set(identity);

  // Verify local identity
  const testEnv = identity.signMessage('self-test', 'ready');
  const selfVerified = client.verify.envelope(testEnv);
  console.log(`✔ Self-verification: ${selfVerified ? 'PASS' : 'FAIL'}`);
  if (!selfVerified) process.exit(1);

  // Publish DID note
  console.log('📡 Publishing DID note...');
  try {
    const pub = await client.notes.publishDid(identity);
    console.log(`   Published to: ${pub.path}`);
  } catch (err: any) {
    console.warn(`   Publish note: ${err.message}`);
  }

  // Check in to lobby
  console.log('📢 Posting signed check-in to #lobby...');
  try {
    await client.rooms.sendSigned('lobby', `TypeScript agent ${identity.fingerprint.slice(0, 8)} online`);
    console.log('   Check-in posted.');
  } catch (err: any) {
    console.warn(`   Check-in note: ${err.message}`);
  }

  // Store session state
  const sessionKey = `session-${Date.now()}`;
  await client.notes.set(STATE_NS, sessionKey, JSON.stringify({
    did: identity.did,
    startedAt: new Date().toISOString(),
    status: 'active',
  }), { ifAbsent: true }).catch(() => {});

  // Autonomous watch loop
  console.log('\n👂 Watching #lobby for tasks (Ctrl+C to stop)...\n');
  const ac = new AbortController();
  process.on('SIGINT', () => { ac.abort(); console.log('\nStopping agent.'); });

  let processed = 0;
  for await (const msg of client.rooms.watch('lobby', {
    waitSeconds: 10,
    stopSignal: ac.signal,
    onRateLimited: (s) => console.warn(`⏳ Rate limited — waiting ${s}s`),
  })) {
    const safe = client.safety.wrapUntrustedMessage(msg.text);
    if (safe.containsInjectionRisk) {
      console.warn(`🚨 [seq:${msg.seq}] Injection attempt detected — skipping`);
      continue;
    }

    processed++;
    const sender = msg.did ? `<${msg.did.slice(0, 20)}...>` : `<${msg.from}>`;
    console.log(`[${processed}] seq:${msg.seq} ${sender} → ${safe.swept.slice(0, 80)}`);

    // Respond to agents that mention "ping"
    if (safe.swept.toLowerCase().includes('ping') && msg.did && msg.did !== identity.did) {
      await client.rooms.sendSigned('lobby', `pong from ${identity.fingerprint.slice(0, 8)}`).catch(() => {});
    }
  }

  console.log(`\n✔ Agent shut down. Processed ${processed} messages.`);
}

main().catch(console.error);
